File: apps/api/src/assessments/import-mapping/llm-client.ts
import Anthropic from '@anthropic-ai/sdk';
import { Logger } from '@nestjs/common';

/**
 * One client per wire format a provider speaks, all satisfying the same narrow interface —
 * `ImportMappingSuggesterService` (and everything below it) only ever calls `complete()` and
 * doesn't care which format or provider answered. Adding a third format (e.g. Google's Gemini
 * API) later is a new class implementing this interface plus a case in `buildClient()` below —
 * nothing else in the suggestion pipeline changes.
 */
export interface LlmClient {
  complete(systemPrompt: string, userPrompt: string): Promise<string>;
}

const REQUEST_TIMEOUT_MS = 15_000;

/** Any OpenAI-compatible `/chat/completions` endpoint — OpenRouter, Groq, Together, a self-hosted vLLM/Ollama server, OpenAI itself. */
export class OpenAiCompatibleClient implements LlmClient {
  constructor(
    private readonly apiKey: string,
    private readonly baseUrl: string,
    private readonly model: string,
  ) {}

  async complete(systemPrompt: string, userPrompt: string): Promise<string> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);

    try {
      const response = await fetch(`${this.baseUrl}/chat/completions`, {
        method: 'POST',
        signal: controller.signal,
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          model: this.model,
          temperature: 0,
          response_format: { type: 'json_object' },
          messages: [
            { role: 'system', content: systemPrompt },
            { role: 'user', content: userPrompt },
          ],
        }),
      });

      if (!response.ok) {
        throw new Error(`LLM endpoint returned ${response.status}: ${await response.text()}`);
      }

      const body = (await response.json()) as { choices?: { message?: { content?: string } }[] };
      const content = body.choices?.[0]?.message?.content;
      if (!content) {
        throw new Error('LLM response had no message content');
      }
      return content;
    } finally {
      clearTimeout(timeout);
    }
  }
}

/**
 * A JSON object of arbitrary string keys/values — deliberately loose, as a plain JSON Schema
 * (not a Zod schema: the SDK's `zodOutputFormat()` helper is typed against zod's newer 'zod/v4'
 * core specifically, a different type identity than the classic 'zod' import this repo uses
 * everywhere else, and pulling that in for one loose schema wasn't worth the friction). The real
 * per-key validation (is this actually one of the unmapped canonical columns? is the value
 * actually one of the file's headers?) happens in `ImportMappingSuggesterService.parseAndValidate`
 * on the raw text either client returns, identically regardless of which one produced it.
 */
const MAPPING_JSON_SCHEMA: Anthropic.Messages.JSONOutputFormat = {
  type: 'json_schema',
  schema: { type: 'object', additionalProperties: { type: 'string' } },
};

/**
 * Anthropic's Messages API, via the official `@anthropic-ai/sdk` (not raw HTTP — Claude doesn't
 * speak the OpenAI wire format `OpenAiCompatibleClient` above targets, and the SDK is the
 * supported way to call it). `output_config.format` constrains the response to the JSON Schema
 * above — Anthropic's equivalent of OpenAI's `response_format: {type: 'json_object'}` — instead
 * of trusting the model to emit valid JSON unprompted.
 */
export class AnthropicMessagesClient implements LlmClient {
  private readonly client: Anthropic;

  constructor(
    apiKey: string,
    private readonly model: string,
    baseUrl?: string,
  ) {
    this.client = new Anthropic({ apiKey, ...(baseUrl ? { baseURL: baseUrl } : {}) });
  }

  async complete(systemPrompt: string, userPrompt: string): Promise<string> {
    const response = await this.client.messages.create({
      model: this.model,
      max_tokens: 1024,
      system: systemPrompt,
      messages: [{ role: 'user', content: userPrompt }],
      output_config: { format: MAPPING_JSON_SCHEMA },
    });

    const textBlock = response.content.find((block): block is Anthropic.Messages.TextBlock => block.type === 'text');
    if (!textBlock) {
      throw new Error(`Claude response had no text content (stop_reason: ${response.stop_reason})`);
    }
    return textBlock.text;
  }
}

/**
 * Tries each configured client in order, only moving to the next on a real failure
 * (error/timeout) — the first one that returns wins. Never blocks: if every provider in the
 * chain fails, the last error propagates to the caller's existing catch-and-degrade-to-empty-
 * mapping handling (see `ImportMappingSuggesterService.suggestMapping`), same as a single
 * misbehaving provider would today.
 */
export class FallbackLlmClient implements LlmClient {
  constructor(private readonly clients: LlmClient[]) {}

  async complete(systemPrompt: string, userPrompt: string): Promise<string> {
    let lastError: unknown;
    for (const client of this.clients) {
      try {
        return await client.complete(systemPrompt, userPrompt);
      } catch (error) {
        lastError = error;
        llmClientLogger.warn(`Provider failed, trying next in the fallback chain: ${error instanceof Error ? error.message : String(error)}`);
      }
    }
    throw lastError;
  }
}

export type ProviderFormat = 'openai' | 'anthropic';

export interface ProviderDefaults {
  format: ProviderFormat;
  baseUrl: string;
  model: string;
}

// Free-tier catalogs churn fast enough that hardcoding a specific model (e.g. a named Hermes
// slot) is a real trap — one such default broke within weeks of being written here. These are
// starting points for an unconfigured slot, each overridable per-field via env vars.
export const SLOT_DEFAULTS: Record<number, ProviderDefaults> = {
  // OpenRouter's own router-level free entry point — picks among whatever's currently free
  // rather than naming one model, so it survives that churn.
  1: { format: 'openai', baseUrl: 'https://openrouter.ai/api/v1', model: 'openrouter/free' },
  // Groq's free tier — no card required, real named open-weight models, fast inference.
  2: { format: 'openai', baseUrl: 'https://api.groq.com/openai/v1', model: 'llama-3.3-70b-versatile' },
  // Paid top-tier fallback — only activates if LLM_PROVIDER_3_API_KEY is actually set.
  3: { format: 'anthropic', baseUrl: 'https://api.anthropic.com', model: 'claude-opus-5' },
};
export const MAX_PROVIDER_SLOTS = 5;

/** Exported so apps/api/src/llm-settings can build the same client types from DB-stored, tenant-owned credentials instead of env vars. */
export function buildClient(format: ProviderFormat, apiKey: string, baseUrl: string, model: string): LlmClient {
  switch (format) {
    case 'anthropic':
      return new AnthropicMessagesClient(apiKey, model, baseUrl);
    case 'openai':
      return new OpenAiCompatibleClient(apiKey, baseUrl, model);
  }
}

/**
 * Reads an ordered chain of providers from `LLM_PROVIDER_<n>_*` env vars (n = 1..5) — each slot
 * needs at minimum `LLM_PROVIDER_<n>_API_KEY` to activate; `_FORMAT`/`_BASE_URL`/`_MODEL` fall
 * back to `SLOT_DEFAULTS` for slots 1-3, or must be set explicitly for slots 4-5. Returns `null`
 * — not a throwing stub — when no slot is configured at all, so every caller's fallback is "skip
 * the suggestion, keep auto-mapping" rather than a startup crash over an optional feature. With
 * exactly one slot configured, returns that client directly (no pointless single-entry chain);
 * with more than one, wraps them in a `FallbackLlmClient` tried in slot order.
 */
export function resolveLlmClient(): LlmClient | null {
  const clients: LlmClient[] = [];

  for (let slot = 1; slot <= MAX_PROVIDER_SLOTS; slot++) {
    const apiKey = process.env[`LLM_PROVIDER_${slot}_API_KEY`];
    if (!apiKey) {
      continue;
    }
    const defaults = SLOT_DEFAULTS[slot];
    const format = (process.env[`LLM_PROVIDER_${slot}_FORMAT`] || defaults?.format) as ProviderFormat | undefined;
    const baseUrl = process.env[`LLM_PROVIDER_${slot}_BASE_URL`] || defaults?.baseUrl;
    const model = process.env[`LLM_PROVIDER_${slot}_MODEL`] || defaults?.model;

    if (!format || !baseUrl || !model) {
      llmClientLogger.warn(`LLM_PROVIDER_${slot}_API_KEY is set but FORMAT/BASE_URL/MODEL has no default for this slot and none was provided — skipping.`);
      continue;
    }
    clients.push(buildClient(format, apiKey, baseUrl, model));
  }

  if (clients.length === 0) {
    return null;
  }
  return clients.length === 1 ? clients[0] : new FallbackLlmClient(clients);
}

export const llmClientLogger = new Logger('ImportMappingSuggester');

File: services/data-analysis/analysis/llm_client.py
"""Provider-agnostic LLM resolution for PandasAI's "ai" analysis mode.

Deliberately mirrors apps/api/src/assessments/import-mapping/llm-client.ts's env var scheme
(`LLM_PROVIDER_<n>_API_KEY` / `_FORMAT` / `_BASE_URL` / `_MODEL`, n = 1..5) and fallback
behaviour (`FallbackLlmClient`), so one set of provider credentials configures both the
TypeScript import-mapping suggester and this service identically. The two are independent
implementations (different languages, different LLM wire clients) but should always agree on
which providers are configured and in what order they're tried.
"""

from __future__ import annotations

import logging
import os
from dataclasses import dataclass
from typing import Optional

from litellm.exceptions import APIError, RateLimitError, ServiceUnavailableError, Timeout
from pandasai.llm.base import LLM
from pandasai_litellm.litellm import LiteLLM

logger = logging.getLogger("data-analysis.llm_client")

MAX_PROVIDER_SLOTS = 5

# Same free-tier-first defaults as the TypeScript side — see llm-client.ts's SLOT_DEFAULTS
# comment for why slot 1 uses OpenRouter's router-level `openrouter/free` rather than a named
# free model (named free models on OpenRouter's catalog churn fast enough that one hardcoded
# here broke within weeks during development).
@dataclass(frozen=True)
class ProviderDefaults:
    format: str
    base_url: Optional[str]
    model: str


SLOT_DEFAULTS: dict[int, ProviderDefaults] = {
    1: ProviderDefaults("openai", "https://openrouter.ai/api/v1", "openrouter/free"),
    2: ProviderDefaults("openai", "https://api.groq.com/openai/v1", "llama-3.3-70b-versatile"),
    3: ProviderDefaults("anthropic", None, "claude-opus-5"),
}

# Failures worth falling through to the next provider on. Deliberately narrower than "any
# Exception" so a bug in our own code (a TypeError from malformed kwargs, say) surfaces instead
# of being silently swallowed and misread as "every provider is down".
_FALLTHROUGH_ERRORS = (APIError, RateLimitError, ServiceUnavailableError, Timeout, ConnectionError)


def _build_litellm(fmt: str, api_key: str, base_url: Optional[str], model: str) -> LiteLLM:
    if fmt == "anthropic":
        return LiteLLM(model=f"anthropic/{model}", api_key=api_key)
    return LiteLLM(model=f"openai/{model}", api_key=api_key, api_base=base_url)


class FallbackLLM(LLM):
    """Tries each configured LiteLLM client in order, only moving to the next on a real
    provider failure — the first one that returns wins. If every provider fails, the last
    error propagates (mirrors FallbackLlmClient.complete in llm-client.ts)."""

    def __init__(self, clients: list[LiteLLM]):
        super().__init__(api_key=None)
        if not clients:
            raise ValueError("FallbackLLM needs at least one underlying client")
        self._clients = clients

    @property
    def type(self) -> str:
        return "fallback"

    def call(self, instruction, context=None) -> str:
        last_error: Exception | None = None
        for client in self._clients:
            try:
                return client.call(instruction, context)
            except _FALLTHROUGH_ERRORS as error:
                last_error = error
                logger.warning("Provider failed, trying next in the fallback chain: %s", error)
        assert last_error is not None
        raise last_error


def resolve_llm(providers_override: Optional[list[dict]] = None) -> Optional[LLM]:
    """Reads LLM_PROVIDER_<n>_* env vars (n = 1..5) into an ordered provider chain. Returns
    None — not a raised error — when no slot is configured, so callers can treat "ai" mode as
    simply unavailable rather than crash the whole service over an optional feature.

    `providers_override`, when given, is used instead of the env vars entirely — this is how a
    tenant's own UI-configured credentials (apps/web/pages/settings/ai.tsx) reach this service:
    apps/api's DataAnalysisService resolves them from the database and forwards the decrypted,
    ordered list over the existing internal HTTP call to this service's /analyze endpoint (see
    main.py's `llm_providers` form field). Each dict needs "format", "apiKey" and "model";
    "baseUrl" is optional (required in practice for "openai", unused for "anthropic").
    """
    if providers_override is not None:
        override_clients: list[LiteLLM] = []
        for provider in providers_override:
            fmt = provider.get("format")
            api_key = provider.get("apiKey")
            model = provider.get("model")
            if not fmt or not api_key or not model:
                logger.warning("Skipping a provider override entry missing format/apiKey/model: %r", provider)
                continue
            override_clients.append(_build_litellm(fmt, api_key, provider.get("baseUrl"), model))
        if not override_clients:
            return None
        return override_clients[0] if len(override_clients) == 1 else FallbackLLM(override_clients)

    clients: list[LiteLLM] = []

    for slot in range(1, MAX_PROVIDER_SLOTS + 1):
        api_key = os.environ.get(f"LLM_PROVIDER_{slot}_API_KEY")
        if not api_key:
            continue

        defaults = SLOT_DEFAULTS.get(slot)
        fmt = os.environ.get(f"LLM_PROVIDER_{slot}_FORMAT") or (defaults.format if defaults else None)
        base_url = os.environ.get(f"LLM_PROVIDER_{slot}_BASE_URL") or (defaults.base_url if defaults else None)
        model = os.environ.get(f"LLM_PROVIDER_{slot}_MODEL") or (defaults.model if defaults else None)

        if not fmt or not model or (fmt == "openai" and not base_url):
            logger.warning(
                "LLM_PROVIDER_%d_API_KEY is set but FORMAT/BASE_URL/MODEL has no default for this slot "
                "and none was provided — skipping.",
                slot,
            )
            continue

        clients.append(_build_litellm(fmt, api_key, base_url, model))

    if not clients:
        return None
    return clients[0] if len(clients) == 1 else FallbackLLM(clients)
